rust/worker/src/execution/operators/merge_metadata_results.rs

```diff
@@ -96,150 +105,327 @@ impl Operator<MergeMetadataResultsOperatorInput, MergeMetadataResultsOperatorOut
+            input.filtered_log.clone(),
+            curr_max_offset_id,
+        );
+        let mat_records = match materializer.materialize().await {
+            Ok(records) => records,
+            Err(e) => {
+                return Err(MergeMetadataResultsOperatorError::LogMaterializationError(
+                    e,
+                ));
             }
         };
 
-        // Hydrate the data from the record segment for filtered data, if
-        // filters were applied
-        if input.filtered_index_offset_ids.is_some() {
-            for index_offset_id in input.filtered_index_offset_ids.as_ref().unwrap().iter() {
-                let record = match record_segment_reader
-                    .get_data_for_offset_id(*index_offset_id as u32)
-                    .await
-                {
-                    Ok(record) => record,
-                    Err(e) => {
-                        println!("Error reading Record Segment: {:?}", e);
-                        return Err(MergeMetadataResultsOperatorError::RecordSegmentReadError);
+        let mut ids: Vec<String> = Vec::new();
+        let mut metadata = Vec::new();
+        let mut documents = Vec::new();
+        let mut visited_ids: HashSet<u32> = HashSet::new();
+        match merged_offset_ids {
+            Some(merged_ids) => {
+                // Hydrate from the logs.
+                for (log, _) in mat_records.iter() {
+                    if merged_ids.contains(&log.offset_id) {
+                        visited_ids.insert(log.offset_id);
+                        if log.final_operation != Operation::Delete {
+                            // Push the final document.
+                            match log.final_document {
+                                Some(doc) => documents.push(Some(doc.to_string())),
+                                None => match log.data_record.as_ref() {
+                                    Some(data_record) => match data_record.document {
+                                        Some(doc) => documents.push(Some(doc.to_string())),
+                                        None => documents.push(None),
+                                    },
+                                    None => documents.push(None),
+                                },
+                            };
+                            // Final metadata.
+                            let mut final_metadata = match log.data_record.as_ref() {
+                                Some(data_record) => match data_record.metadata {
+                                    Some(ref map) => map.clone(), // auto deref here.
+                                    None => HashMap::new(),
+                                },
+                                None => HashMap::new(),
+                            };
+                            if log.metadata_to_be_merged.as_ref().is_some() {
+                                for (key, value) in log.metadata_to_be_merged.as_ref().unwrap() {
```

Comment:
should we make merge logic reused

Reply:
Nones are not handled yet. It's a separate task that we created today in our tracker